azure-iot-device/tests/iothub/auth/test_iotedge_authentication_provider.py

```diff
@@ -4,61 +4,318 @@
 # license information.
 # --------------------------------------------------------------------------
 
-import os
+import pytest
+import requests
+import json
+import base64
+import six.moves.urllib as urllib
 from azure.iot.device.iothub.auth.iotedge_authentication_provider import (
-    IotEdgeAuthenticationProvider,
+    IoTEdgeAuthenticationProvider,
+    IoTEdgeHsm,
+    IoTEdgeError,
 )
-from mock import patch
-
-fake_ca_cert = "__FAKE_CA_CERTIFICATE__"
-fake_module_id = "__FAKE_MODULE_ID__"
-fake_device_id = "__FAKE_DEVICE_ID__"
-fake_hostname = "__FAKE_HOSTNAME__"
-fake_gateway_hostname = "__FAKE_GATEWAY_HOSTNAME__"
-fake_digest = "__FAKE_DIGEST__"
-
-required_environment_variables = {
-    "IOTEDGE_MODULEID": fake_module_id,
-    "IOTEDGE_DEVICEID": fake_device_id,
-    "IOTEDGE_IOTHUBHOSTNAME": fake_hostname,
-    "IOTEDGE_GATEWAYHOSTNAME": fake_gateway_hostname,
-}
-
-
-@patch.dict(os.environ, required_environment_variables)
-@patch("azure.iot.device.iothub.auth.iotedge_authentication_provider.IotEdgeHsm")
-def test_initializer_gets_details_from_environment(mock_hsm):
-    auth_provider = IotEdgeAuthenticationProvider()
-    assert auth_provider.gateway_hostname == fake_gateway_hostname
-    assert auth_provider.device_id == fake_device_id
-    assert auth_provider.module_id == fake_module_id
-    assert auth_provider.hostname == fake_hostname
-
-
-@patch.dict(os.environ, required_environment_variables)
-@patch("azure.iot.device.iothub.auth.iotedge_authentication_provider.IotEdgeHsm")
-def test_initializer_gets_ca_certificate_from_hsm(MockHsm):
-    MockHsm.return_value.get_trust_bundle.return_value = fake_ca_cert
-    auth_provider = IotEdgeAuthenticationProvider()
-    assert auth_provider.ca_cert == fake_ca_cert
-
-
-@patch.dict(os.environ, required_environment_variables)
-@patch("azure.iot.device.iothub.auth.iotedge_authentication_provider.IotEdgeHsm")
-def test_get_shared_access_key_uses_hsm_to_sign(MockHsm):
-    MockHsm.return_value.sign.return_value = fake_digest
-    auth_provider = IotEdgeAuthenticationProvider()
-    try:
-        sas_token = auth_provider.get_current_sas_token()
-        assert MockHsm.return_value.sign.call_args[0][0].startswith(
-            "{}%2Fdevices%2F{}%2Fmodules%2F{}\n".format(
-                fake_hostname, fake_device_id, fake_module_id
-            )
+from .shared_auth_tests import SharedBaseRenewableAuthenticationProviderInstantiationTests
+
+
+@pytest.fixture
+def gateway_hostname():
+    return "__FAKE_GATEWAY_HOSTNAME__"
+
+
+@pytest.fixture
+def module_generation_id():
+    return "__FAKE_MODULE_GENERATION_ID__"
```

Comment:
Interesting.  Offline, I'd like to hear your opinion of this (using fixtures for simple strings that hold fake values).  Do you think this is a reasonable practice and worth the overhead (if it is indeed overhead)?

Reply:
For the simple cases like this I feel literal values of string or integer would do.

---
In reply to: [294473417](https://github.com/Azure/azure-iot-sdk-python-preview/pull/110#discussion_r294473417) [](ancestors = 294473417)

Reply:
Because these values have to be shared between multiple test modules, not just for various things that all use these values, but also for extensions of inherited methods, I believe fixtures are the best way to do it. Not only are they easily shared between tests modules without imports, but they can fit easily be used across an inheritance hierarchy while keeping the signatures the same. If that makes sense. 

I'll likely have to explain this in a bit more detail with additional clarity later, but this technique is crucial for testing complex object hierarchies.

---
In reply to: [294500681](https://github.com/Azure/azure-iot-sdk-python-preview/pull/110#discussion_r294500681) [](ancestors = 294500681,294473417)